### DISCUSSION TITLE
How do I find the owner of Taproot Assets?

### DISCUSSION THREAD

**Discussion by fanshome1001** [lightninglabs/taproot-assets#809] [Q&A]
What method should I use to find out which addresses are holding an asset with a known asset id?

**Comment by guggero** [comment]
For Taproot Asset Protocol addresses created by your `tapd` you can simply run `tapcli addrs query` and check the `asset_id` in the response. There is no way to directly filter the output by asset ID yet.
The corresponding RPC method is [`QueryAddrs`](https://lightning.engineering/api-docs/api/taproot-assets/taproot-assets/query-addrs).

**Comment by fanshome1001** [comment]
The above method does not solve my problem. I guess it would take reading sparse Merkle sum trees to count the holders of a given asset id, but how?

**Reply by guggero** [reply]
Maybe you could outline a bit more what exactly you want to know? When you say "addresses", do you mean Taproot Asset addresses (e.g. starting with `tapbc1...`) or on-chain Bitcoin addresses that commit assets?
And when you say "holders of a given asset", do you mean within your node or globally?

**Comment by fanshome1001** [comment]
I would like to know what are the Bitcoin addresses on the chain that hold an asset within globally, and how many are held at each address?

**Reply by guggero** [reply]
Okay, I understand. Because Taproot Assets is fully off-chain and privacy focused, this information is not available globally, at least not per default.
The information what assets are minted and transferred in what on-chain transactions is only contained in the proof files. Those proofs _can_ be uploaded to a universe server, **if the user chooses to do that**.
Lightning Labs runs such a universe server and `tapd` by default uses that for issuance and transfer proofs. But users can run their own (public or private) universe servers if they choose to do that.

Assuming you want to retrieve the proofs that have been uploaded to a certain universe, you can use the [REST API](https://lightning.engineering/api-docs/category/universe-service) to retrieve the information.
Here is an example with the public universe server from Lightning Labs (on testnet):

1. List all assets known to that universe: https://universe.testnet.lightningcluster.com/v1/taproot-assets/universe/stats/assets?sort_by=SORT_BY_ASSET_ID
2. List all issuance proofs for a (randomly picked) asset: https://universe.testnet.lightningcluster.com/v1/taproot-assets/universe/keys/asset-id/f1d1c27d04f207f82b3462da9d242f890ac8bc1ea7bfae8f10ca5a90995be286?id.proof_type=PROOF_TYPE_ISSUANCE
3. The `op_str` field above is where on chain that asset is located. Looking it up in a block explorer would correspond to this transaction (output with index 0): https://blockstream.info/testnet/tx/77dc414d73b4acbb91a032ee4db3a9acbb344baf4f80376a8a31aef50c8ccc14?expand

**Comment by fanshome1001** [comment]
Thank you. In another scenario, if an asset issuer needs to disclose the distribution of the asset, what should he do? How the user should upload the proof files to a universe server ,if the user chooses to do that. and how others or asset issuers should download them from the universe server.

**Reply by fanshome1001** [reply]
When I use "universe proofs insert", I get an error: [tapcli] rpc error: code = Unknown desc = proof insert is disabled for the given universe. I checked the input parameters carefully and found no problem.

**Reply by guggero** [reply]
See https://docs.lightning.engineering/lightning-network-tools/taproot-assets/universes, that should explain things to you.
